connection/connection.go

```diff
@@ -70,12 +72,11 @@ func SetMaxGRPCLogLength(characterCount int) {
 // For other connections, the default behavior from gRPC is used and
 // loss of connection is not detected reliably.
 func Connect(address string, metricsManager metrics.CSIMetricsManager, options ...Option) (*grpc.ClientConn, error) {
-	return connect(address, metricsManager, []grpc.DialOption{grpc.WithTimeout(time.Second * 30)}, options)
-}
-
-// ConnectWithoutMetrics behaves exactly like Connect except no metrics are recorded.
-func ConnectWithoutMetrics(address string, options ...Option) (*grpc.ClientConn, error) {
-	return connect(address, nil, []grpc.DialOption{grpc.WithTimeout(time.Second * 30)}, options)
+	options = append(options, withTimeout(time.Second*30))
+	if metricsManager != nil {
+		options = append(options, withMetrics(metricsManager))
+	}
+	return connect(address, options)
 }
```

Comment:
I was thinking, a builder pattern might be more suitable for building options to connect. So something like:

```golang
type ConnectionBuilder struct {
	address           string
	metricsManager    metrics.CSIMetricsManager
	timeout           time.Duration
	enableOtelTracing bool
}

func NewConnectionBuilder(address string) *ConnectionBuilder {
	return &ConnectionBuilder{
		address: address,
	}
}

func (b *ConnectionBuilder) WithMetrics(mm metrics.CSIMetricsManager) *ConnectionBuilder {
	b.metricsManager = mm
	return b
}

func (b *ConnectionBuilder) WithTimeout(td time.Duration) *ConnectionBuilder {
	b.timeout = td
	return b
}

func (b *ConnectionBuilder) WithOtelTracing() *ConnectionBuilder {
	b.enableOtelTracing = true
	return b
}

func (b *ConnectionBuilder) Connect() (*grpc.ClientConn, error) {
	return connect(b.address, b.metricsManager,....)
}
```

So a caller would do something like:

```
connectionBuilder := NewConnectionBuilder("unix:////blah").WithMetrics(xxx).WithTimeout(xxx)
connection, err := connectionBuilder.Connect()
```

This might be more easier to extend in future. I do realize, this is new interface functions but it should be fine to use it. We can eventually deprecate both `Connect` and `ConnectWithoutMetrics` in future.

What do you think?

Reply:
I do not have a strong opinion on this but here are some thoughts:
- I am not a big fan of breaking the interface because it means people will have to update their code even if they do not want to change the behaviour of `Connect`. Considering `csi-lib-utils` is used in all sidecars and some csi-drivers that makes a fair amount of 'clients' that will have to update their code.
- initially, I sticked to the `Options` pattern because I saw a similar patterns in other CSI repos ([metrics in csi-lib-utils](https://github.com/kubernetes-csi/external-resizer/blob/master/vendor/github.com/kubernetes-csi/csi-lib-utils/metrics/metrics.go#L222), [aws-ebs-csi-driver](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/master/pkg/driver/driver.go#L76), [azuredisk-csi-driver](https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/master/pkg/azuredisk/azuredisk.go#L138), ...) so I felt like sticking to the same pattern was a good idea for code consistency.
- on the other hand, I did not see any other builder pattern in CSI repos (if you have examples, feel free to share them) so it does not feel very consistent with the existing code.

But if you feel like a builder pattern is truly the way to go here, I'm fine with implementing it.
Also, maybe we can ask other folks' opinion on this? Not too familiar with who would be the right person though :/

Reply:
okay. I can probably live with function(option) pattern.  I have a minor question below. Another thing  is - we are about to start cutting CSI sidecars. Were you hoping to merge this in sidecars and start adding CLI option?  Do you want to do this after this release or in current release? (I pinged you on slack with same question).

Reply:
Btw supplying a variadic number of arguments for setting options vs builder pattern are basically two sides of the same coin and not all that much different.  So basically:

```golang
foo = Connect("foo", withBar(), withBaz(), withFoo())

vs

foo = NewConnect("foo").WithBar().WithBaz().WithFoo()

```

I am not a fan of functions that takes too many parameters. Also it may just be me, but those functions that returns functions, only for setting a field in a option struct seems slightly complicated for its own good (it takes a minute for me to parse those), where as builder seems more straightforward.

I agree that we shouldn't break existing interfaces and we can leave `Connect` as it is, but I do tend to prefer the API of builder vs functions that takes options.